proposals/0165-async-vote-execution.md

```diff
@@ -0,0 +1,218 @@
+epoch.
+
+### Replay the full block on selected forks asynchronously
+
+There is no protocol enforced order of block replay for various validator
+implemenations, new vote transactions could be sent when the vote only hash
+or replay tip hash changes.
+
+Once a validator has determined the fork it will vote on, it can prioritize
+replaying blocks on the selected fork. The replay process is the same as today,
+all transactions (vote and non-vote) will be executed to determine the final
+bankhash. The computed bankhash will be attached to vote instructions. So we
+can still detect non-determinism (same of set of instructions leading to
+different results) like today, only that maybe at a later time.
+
+To guarantee the blockchain will halt when full replay is having problems, we
+propose:
+
+1. If full replay is behind vote only replay by more than 1/2 epoch or vice
+versa, stop producing new blocks until the lagging replay catches up. Also set
+up monitoring if the distance between two replays are growing larger.
+
+2. If more than 1/3 of the validators send a different final hash for a block
+with the same vote only hash, panic and prompt for further debugging.
+
+In this step the validators will check the fee payers of vote transactions. So
+each vote transaction is executed twice: once in the optimistic voting stage
+*without* checking fee payer, and once in this stage *with* checking fee payer.
+If a staked validator does not have vote fee covered for specific votes, we
+will not accept the vote today, while in the future we accept the vote in fork
+selection, but does not actually give vote credits because the transaction
+failed.
+
+### Enable Async Vote Executions
+
+1. The leader will no longer execute any transactions before broadcasting
+the block it packed. We do have block-id (Merkle tree root) to ensure
+everyone receives the same block.
+2. Upon receiving a new block, the validator executes only the vote
+transactions without checking the fee payers. The result is immediately
+applied in consensus to select a fork. Then votes are sent out for the
+selected fork with the `Vote only bankhash` for the tip of the fork and the
+most recent `Replay tip bankhash`.
+3. The blocks on the selected forks are scheduled to be replayed. When
+a block is replayed, all transactions are executed with fee payers checked.
+This is the same as the replay we use today.
+4. Optimisticly confirmed or finalized on `Vote only bankhash` and `Replay tip
+bankhash` will be both exposed through RPC so users can select accordingly.
+5. Add assertion that confirmed `Replay tip bankhash` is not too far away from
+the confirmed `Vote only bankhash` (currently proposed at 1/2 of the Epoch)
+6. Add alerts if `Replay tip bankhash` differs when the `Vote only bankhash` is
+the same. This is potentially an event worthy of cluster restart. If more than
+1/3 of the validators claim a different `Replay tip bankhash`, halt and exit.
+
```

Comment:
Specify the modifications to fork choice. Fork choice will now *only* read vote only replayed votes. Needs to be keyed by block-id or vote only hash instead.

Reply:
It's covered in "The result is immediately applied in consensus to select a fork." I think whether we key our internal data structure by block-id or hash is implementation details we don't need in SIMD. I've added that to point 2 in "Enable Async Vote Executions".